distributed/dashboard/components/scheduler.py

```diff
@@ -0,0 +1,1797 @@
+            self.last = 0
+            self.scheduler = scheduler
+            self.source = ColumnDataSource(
+                {
+                    "bandwidth": [1, 2],
+                    "source": ["a", "b"],
+                    "destination": ["a", "b"],
+                    "bandwidth_text": ["1", "2"],
+                }
+            )
+
+            values = [hex(x)[2:] for x in range(64, 256)][::-1]
+            mapper = linear_cmap(
+                field_name="bandwidth",
+                palette=["#" + x + x + "FF" for x in values],
+                low=0,
+                high=1,
+            )
+
+            fig = figure(
+                title="Bandwidth by Worker",
+                tools="",
+                id="bk-bandwidth-worker-plot",
+                name="bandwidth_worker_heatmap",
+                x_range=["a", "b"],
+                y_range=["a", "b"],
+                **kwargs
+            )
+            fig.xaxis.major_label_orientation = -math.pi / 12
+            rect = fig.rect(
+                source=self.source,
+                x="source",
+                y="destination",
+                color=mapper,
+                height=1,
+                width=1,
+            )
+
+            self.color_map = mapper["transform"]
+            color_bar = ColorBar(
+                color_mapper=self.color_map,
+                label_standoff=12,
+                border_line_color=None,
+                location=(0, 0),
+            )
+            color_bar.formatter = NumeralTickFormatter(format="0 b")
+            fig.add_layout(color_bar, "right")
+
+            fig.toolbar.logo = None
+            fig.toolbar_location = None
+
+            hover = HoverTool()
+            hover.tooltips = """
+            <div>
+                <p><b>Source:</b> @source </p>
+                <p><b>Destination:</b> @destination </p>
+                <p><b>Bandwidth:</b> @bandwidth_text / s</p>
+            </div>
+            """
+            hover.point_policy = "follow_mouse"
+            fig.add_tools(hover)
+
+            self.fig = fig
+
+    @without_property_validation
+    def update(self):
+        with log_errors():
+            bw = self.scheduler.bandwidth_workers
+            if not bw:
+                return
+            x, y, value = zip(*[(a, b, c) for (a, b), c in bw.items()])
+
+            if self.color_map.high < max(value):
+                self.color_map.high = max(value)
+
+            factors = list(sorted(set(x + y)))
+            self.fig.x_range.factors = factors
+            self.fig.y_range.factors = factors
+
+            result = {
+                "source": x,
+                "destination": y,
+                "bandwidth": value,
+                "bandwidth_text": list(map(format_bytes, value)),
+            }
+            self.fig.title.text = "Bandwidth: " + format_bytes(self.scheduler.bandwidth)
+
+            update(self.source, result)
+
+
+class CurrentLoad(DashboardComponent):
```

Comment:
This should maybe be broken out into one class per plot?  I'm not sure.

Reply:
Possibly. The nvml stuff also seems to follow this pattern.

Reply:
I've had a go at pulling this apart into two classes and ended up with quite a bit of duplication. I'm tempted to say that while this class is quite big and complex it does make sense to gather these metrics in one go.

An alternative would be to make a common base class which deals with the data updating and then subclasses which deal with plotting the parts they want. However, that introduces quite a lot of indirection.

What do you think @mrocklin?

Reply:
Sounds fine to me